File: Android/java/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/network/core/MimeMatcher.java
/*
 * Copyright (c) 2014-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

package com.didichuxing.doraemonkit.kit.network.core;

import android.annotation.SuppressLint;

import java.util.ArrayList;

public class MimeMatcher<T> {
  private final ArrayList<MimeMatcherRule> mRuleMap = new ArrayList<MimeMatcherRule>();

  public void addRule(String ruleExpression, T resultIfMatched) {
    mRuleMap.add(new MimeMatcherRule(ruleExpression, resultIfMatched));
  }

  public void clear() {
    mRuleMap.clear();
  }

  public T match(String mimeT) {
    int ruleMapN = mRuleMap.size();
    for (int i = 0; i < ruleMapN; i++) {
      MimeMatcherRule rule = mRuleMap.get(i);
      if (rule.match(mimeT)) {
        return rule.getResultIfMatched();
      }
    }
    return null;
  }

  @SuppressLint("BadMethodUse-java.lang.String.length")
  private class MimeMatcherRule {
    private final boolean mHasWildcard;
    private final String mMatchPrefix;
    private final T mResultIfMatched;

    public MimeMatcherRule(String ruleExpression, T resultIfMatched) {
      if (ruleExpression.endsWith("*")) {
        mHasWildcard = true;
        mMatchPrefix = ruleExpression.substring(0, ruleExpression.length() - 1);
      } else {
        mHasWildcard = false;
        mMatchPrefix = ruleExpression;
      }
      if (mMatchPrefix.contains("*")) {
        throw new IllegalArgumentException("Multiple wildcards present in rule expression " +
            ruleExpression);
      }
      mResultIfMatched = resultIfMatched;
    }

    public boolean match(String mimeType) {
      if (!mimeType.startsWith(mMatchPrefix)) {
        return false;
      }
      return (mHasWildcard || mimeType.length() == mMatchPrefix.length());
    }

    public T getResultIfMatched() {
      return mResultIfMatched;
    }
  }
}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/network/okhttp/core/MimeMatcher.kt
/*
 * Copyright (c) 2014-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
package com.didichuxing.doraemonkit.kit.network.okhttp.core

import android.annotation.SuppressLint
import java.util.*

class MimeMatcher<T> {
    private val mRuleMap = ArrayList<MimeMatcherRule>()
    fun addRule(ruleExpression: String?, resultIfMatched: T) {
        mRuleMap.add(MimeMatcherRule(ruleExpression!!, resultIfMatched))
    }

    fun clear() {
        mRuleMap.clear()
    }

    fun match(mimeT: String?): T? {
        val ruleMapN = mRuleMap.size
        for (i in 0 until ruleMapN) {
            val rule = mRuleMap[i]
            if (rule.match(mimeT!!)) {
                return rule.resultIfMatched
            }
        }
        return null
    }

    @SuppressLint("BadMethodUse-java.lang.String.length")
    private inner class MimeMatcherRule(ruleExpression: String, resultIfMatched: T) {
        private var mHasWildcard = false
        private lateinit var mMatchPrefix: String
        val resultIfMatched: T
        fun match(mimeType: String): Boolean {
            return if (!mimeType.startsWith(mMatchPrefix)) {
                false
            } else mHasWildcard || mimeType.length == mMatchPrefix.length
        }


        init {
            if (ruleExpression.endsWith("*")) {
                mHasWildcard = true
                mMatchPrefix = ruleExpression.substring(0, ruleExpression.length - 1)
            } else {
                mHasWildcard = false
                mMatchPrefix = ruleExpression
            }
            require(!mMatchPrefix.contains("*")) {
                "Multiple wildcards present in rule expression " +
                        ruleExpression
            }
            this.resultIfMatched = resultIfMatched
        }
    }
}